x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteJobAction.java

```diff
@@ -113,58 +125,116 @@ protected AcknowledgedResponse newResponse() {
         return new AcknowledgedResponse();
     }
 
+    @Override
+    protected ClusterBlockException checkBlock(DeleteJobAction.Request request, ClusterState state) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
+    }
+
+    @Override
+    protected void masterOperation(DeleteJobAction.Request request, ClusterState state, ActionListener<AcknowledgedResponse> listener) {
+        throw new UnsupportedOperationException("the Task parameter is required");
+    }
+
     @Override
     protected void masterOperation(Task task, DeleteJobAction.Request request, ClusterState state,
                                    ActionListener<AcknowledgedResponse> listener) {
+        logger.debug("Deleting job '{}'", request.getJobId());
+
+        TaskId taskId = new TaskId(clusterService.localNode().getId(), task.getId());
+        ParentTaskAssigningClient parentTaskClient = new ParentTaskAssigningClient(client, taskId);
+
+        // Check if there is a deletion task for this job already and if yes wait for it to complete
+        Optional<Task> existingStartedTask;
```

Comment:
I think with this change we no longer have a race condition, when two tasks could end up waiting for each other as we had before, but it is still very complicated (especially the waitForExistingTaskToComplete part). What would you think about having a static map of arrays of listeners here. On start, in a critical section, you can check if there is already an element with the same job ID present in the map if not, you add your job ID with an empty array and start delete process, if job ID exists - just add yourself to the array of listeners. On completion, again in a critical section, you remove the array with your job ID from the map and execute all listeners that were registered there.

Reply:
Happy to try that! It sounds less complex indeed. Why do you suggest the listener-map should be static? Given the action is a singleton object, it could just be a final member, right?

Reply:
> Why do you suggest the listener-map should be static?

Well, not enough coffee, I guess. Yes, it should be just a member of action since action is a singleton.

Reply:
I have pushed a commit to refactor using listeners. It did simplify things a lot. Could you have another look please?